onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -101,37 +105,79 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
         {0, i, "Unsqueeze", {1, 11}, kOnnxDomain},
         {0, 0, "Gather", {1, 11}, kOnnxDomain},
         {0, 0, "Shape", {1}, kOnnxDomain}};
-
     std::vector<const Node::EdgeEnd*> edges;
-    if (!graph_utils::FindPath(concat, true, parent_path, edges, logger)) {
-      return false;
-    }
-
-    const Node& unsqueeze = edges[0]->GetNode();
-    const Node& gather = edges[1]->GetNode();
-    const Node& shape = edges[2]->GetNode();
-
-    const NodeArg& shape_input = *(shape.InputDefs()[0]);
-    if (shape_input.Name() != root_input.Name()) {
-      return false;
-    }
-
-    std::vector<int64_t> axes;
-    if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
-      return false;
+    if (graph_utils::FindPath(concat, true, parent_path, edges, logger)) {
+      const Node& unsqueeze = edges[0]->GetNode();
+      const Node& gather = edges[1]->GetNode();
+      const Node& shape = edges[2]->GetNode();
+
+      const NodeArg& shape_input = *(shape.InputDefs()[0]);
+      if (shape_input.Name() != root_input.Name()) {
+        return false;
+      }
+
+      std::vector<int64_t> axes;
+      if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
+        return false;
+      }
+
+      if (!optimizer_utils::IsInitializerWithExpectedValue(graph, *(gather.InputDefs()[1]), int64_t(shape_value.size()), false)) {
+        return false;
+      }
+
+      shape_value.push_back(0);
+
+      candidates_for_removal.push_back(unsqueeze.Index());
+      candidates_for_removal.push_back(gather.Index());
+      candidates_for_removal.push_back(shape.Index()); 
+    } else {
+      // Find subgraph
+      if (subgraph_cnt > 0 || std::find(shape_value.begin(), shape_value.end(), -1) != shape_value.end()) {
+        // Only one subgraph can be used to fuse, and each shape node should contain only one value of -1. 
+        return false;
+      }
+      Node* p_cur_node = const_cast<Node*>(graph_utils::GetInputNode(concat, i));
+      if (p_cur_node == nullptr) {
+        return false;
+      }
+      subgraph_candidates.clear();
+      // From the current node, find the subgraph bottom-up util it reaches the root input node. 
```

Comment:
We can probably add a utility function that remove all nodes that has no any output edges. Actually, there is no need to check whether it reaches the root input node.

It is similar to prune_graph in the python code.

Reply:
Keep in mind there's no edge for a graph output, so just because a node has no output edges doesn't necessarily mean it can be removed.

---
In reply to: [426868760](https://github.com/microsoft/onnxruntime/pull/3931#discussion_r426868760) [](ancestors = 426868760)